src/backends/kubernetes_backend.py

```diff
@@ -0,0 +1,353 @@
+from warnet.utils import parse_raw_messages, default_bitcoin_conf_args
+
+DOCKER_REGISTRY_CORE = "bitcoindevproject/k8s-bitcoin-core"
+DOCKER_REGISTRY_LND = "lightninglabs/lnd:v0.17.0-beta"
+POD_PREFIX = "tank"
+BITCOIN_CONTAINER_NAME = "bitcoin"
+LN_CONTAINER_NAME = "ln"
+
+class KubernetesBackend(BackendInterface):
+
+    def __init__(self, config_dir: Path, network_name: str, namespace = "default", logs_pod = "fluentd") -> None:
+        super().__init__(config_dir)
+        # assumes the warnet rpc server is always
+        # running inside a k8s cluster as a statefulset
+        config.load_incluster_config()
+        self.client = client.CoreV1Api()
+        self.namespace = namespace
+        self.logs_pod = logs_pod
+        self.network_name = network_name
+
+    def build(self) -> bool:
+        # TODO: just return true for now, this is so we can be running either docker or k8s as a backend
+        # on the same branch
+        return True
+
+    def up(self, warnet) -> bool:
+        self.deploy_pods(warnet)
+        return True
+
+    def down(self, warnet) -> bool:
+        """
+        Bring an exsiting network down.
+            e.g. `k delete -f warnet-tanks.yaml`
+        """
+        for tank in warnet.tanks:
+            self.client.delete_namespaced_pod(tank.container_name, self.namespace)
+    
+    def get_file(self, tank_index: int, service: ServiceType, file_path: str):
+        """
+        Read a file from inside a container
+        """
+        pod_name = self.get_pod_name(tank_index, service)
+        exec_command = ['cat', file_path]
+
+        resp = stream(self.client.connect_get_namespaced_pod_exec, pod_name, self.namespace,
+                      command=exec_command,
+                      stderr=True, stdin=True,
+                      stdout=True, tty=False,
+                      _preload_content=False,
+                      container=BITCOIN_CONTAINER_NAME if service == ServiceType.BITCOIN else LN_CONTAINER_NAME)
+
+        file = io.BytesIO()
+        while resp.is_open():
+            resp.update(timeout=1)
+            if resp.peek_stdout():
+                file.write(resp.read_stdout())
+            if resp.peek_stderr():
+                raise Exception("Problem copying file from pod" + resp.read_stderr().decode('utf-8'))
+
+        return file.getvalue()
+
+    def get_pod_name(self, tank_index: int) -> str:
+        return f"{self.network_name}-{POD_PREFIX}-{tank_index:06d}"
+
+    def get_pod(self, container_name: str) -> V1Pod:
+        return cast(V1Pod, self.client.read_namespaced_pod(name=container_name, namespace="default"))
+
+    # We could enhance this by checking the pod status as well
+    # The following pod phases are available: Pending, Running, Succeeded, Failed, Unknown
+    # For example not able to pull image will be a phase of Pending, but the container status will be ErrImagePull
+    def get_status(self, tank_index: int, service: ServiceType) -> RunningStatus:
```

Comment:
`warcli network status` error: `TypeError: Object of type V1PodStatus is not JSON serializable`

Stack:

```
2023-12-01 19:54:24,091 - warnet.server - ERROR - Exception on /api [POST]
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2190, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1486, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1484, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 1469, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/usr/local/lib/python3.9/site-packages/flask/views.py", line 109, in view
    return current_app.ensure_sync(self.dispatch_request)(**kwargs)
  File "/usr/local/lib/python3.9/site-packages/flask/views.py", line 190, in dispatch_request
    return current_app.ensure_sync(meth)(**kwargs)
  File "/usr/local/lib/python3.9/site-packages/flask_jsonrpc/views.py", line 49, in post
    return make_response(jsonify(response), status_code, headers)
  File "/usr/local/lib/python3.9/site-packages/flask/json/__init__.py", line 170, in jsonify
    return current_app.json.response(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/flask/json/provider.py", line 215, in response
    f"{self.dumps(obj, **dump_args)}\n", mimetype=self.mimetype
  File "/usr/local/lib/python3.9/site-packages/flask/json/provider.py", line 180, in dumps
    return json.dumps(obj, **kwargs)
  File "/usr/local/lib/python3.9/json/__init__.py", line 234, in dumps
    return cls(
  File "/usr/local/lib/python3.9/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/local/lib/python3.9/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/local/lib/python3.9/site-packages/flask/json/provider.py", line 120, in _default
    raise TypeError(f"Object of type {type(o).__name__} is not JSON serializable")
TypeError: Object of type V1PodStatus is not JSON serializable
```

Reply:
Thanks, got this and a few more other bits on the k8s backend that aren't currently working or complete. `get_file` is untested and build custom image isn't even past concept stage.

Reply:
I think these serializable errors can be fixed with a `cast` argument, btw. I seem to remember doing that in a few places where the errors were annoying me